File: android/capacitor/src/main/java/com/getcapacitor/PluginConfig.kt
package com.getcapacitor

import com.getcapacitor.util.JSONUtils
import org.json.JSONObject

/**
 * Represents the configuration options for plugins used by Capacitor
 *
 * @param configJSON A plugin configuration expressed as a JSON Object
 */
public class PluginConfig internal constructor(
    /**
     * The JSON Object containing the config of the the provided plugin ID.
     */
    public val configJSON: JSONObject,
) {
    /**
     * Get a string value for a plugin in the Capacitor config.
     *
     * @param configKey The key of the value to retrieve
     * @param defaultValue A default value to return if the key does not exist in the config
     * @return The value from the config, if key exists. Default value returned if not
     */
    public fun getString(configKey: String, defaultValue: String? = null): String? = JSONUtils.getString(configJSON, configKey, defaultValue)

    /**
     * Get a boolean value for a plugin in the Capacitor config.
     *
     * @param configKey The key of the value to retrieve
     * @param defaultValue A default value to return if the key does not exist in the config
     * @return The value from the config, if key exists. Default value returned if not
     */
    public fun getBoolean(configKey: String, defaultValue: Boolean): Boolean = JSONUtils.getBoolean(configJSON, configKey, defaultValue)

    /**
     * Get an integer value for a plugin in the Capacitor config.
     *
     * @param configKey The key of the value to retrieve
     * @param defaultValue A default value to return if the key does not exist in the config
     * @return The value from the config, if key exists. Default value returned if not
     */
    public fun getInt(configKey: String, defaultValue: Int): Int = JSONUtils.getInt(configJSON, configKey, defaultValue)

    /**
     * Get a double value for a plugin in the Capacitor config.
     *
     * @param configKey The key of the value to retrieve
     * @param defaultValue A default value to return if the key does not exist in the config
     * @return The value from the config, if key exists. Default value returned if not
     */
    public fun getDouble(configKey: String, defaultValue: Double): Double = JSONUtils.getDouble(configJSON, configKey, defaultValue)

    /**
     * Get a string array value for a plugin in the Capacitor config.
     *
     * @param configKey The key of the value to retrieve
     * @param defaultValue A default value to return if the key does not exist in the config
     * @return The value from the config, if key exists. Default value returned if not
     */
    public fun getArray(configKey: String, defaultValue: Array<String>? = null): Array<String>? =
        JSONUtils.getArray(configJSON, configKey, defaultValue)

    /**
     * Get a JSON object value for a plugin in the Capacitor config.
     *
     * @param configKey The key of the value to retrieve
     * @return The value from the config, if exists. Null if not
     */
    public fun getObject(configKey: String): JSONObject? = JSONUtils.getObject(configJSON, configKey)

    /**
     * Check if the PluginConfig is empty.
     *
     * @return true if the plugin config has no entries
     */
    public fun isEmpty(): Boolean = configJSON.length() == 0
}

File: ios/Capacitor/Capacitor/PluginConfig.swift
import Foundation

/// The configuration of one plugin: the object under `plugins.<jsName>` in `capacitor.config`, as ``CAPPlugin/getConfig()``
/// returns it.
///
/// Keys may be key paths such as `"ios.style"`. Read single values with the getters, or decode the whole object into a
/// `Decodable` type with ``decode(_:decoder:)``:
///
/// ```swift
/// struct SplashConfig: Decodable {
///     var launchShowDuration: Int?
///     var backgroundColor: String?
/// }
///
/// let config = try getConfig().decode(SplashConfig.self)
/// let duration = config.launchShowDuration ?? 500
/// ```
public struct PluginConfig {

    // The object containing the plugin config values
    private let config: JSObject

    init(config: JSObject) {
        self.config = config
    }

    public func getString(_ configKey: String, _ defaultValue: String? = nil) -> String? {
        if let val = (self.config)[keyPath: KeyPath(configKey)] as? String {
            return val
        }
        return defaultValue
    }

    public func getBoolean(_ configKey: String, _ defaultValue: Bool) -> Bool {
        if let val = (self.config)[keyPath: KeyPath(configKey)] as? Bool {
            return val
        }
        return defaultValue
    }

    public func getInt(_ configKey: String, _ defaultValue: Int) -> Int {
        if let val = (self.config)[keyPath: KeyPath(configKey)] as? Int {
            return val
        }
        return defaultValue
    }

    public func getDouble(_ configKey: String, _ defaultValue: Double) -> Double {
        if let val = (self.config)[keyPath: KeyPath(configKey)] as? Double {
            return val
        }
        return defaultValue
    }

    public func getArray(_ configKey: String, _ defaultValue: JSArray? = nil) -> JSArray? {
        if let val = (self.config)[keyPath: KeyPath(configKey)] as? JSArray {
            return val
        }
        return defaultValue
    }

    public func getObject(_ configKey: String) -> JSObject? {
        return (self.config)[keyPath: KeyPath(configKey)] as? JSObject
    }

    public func isEmpty() -> Bool {
        return self.config.isEmpty
    }

    /**
     * Gets the JSObject containing the config of the the provided plugin ID.
     *
     * @return The config for that plugin
     */
    public func getConfigJSON() -> JSObject {
        return self.config
    }

    /// Decodes the whole configuration into `type`.
    ///
    /// - Parameters:
    ///   - type: The type to decode. Keys missing from the configuration need optional properties: a synthesized
    ///     `Decodable` conformance requires the key of every other property, even one with a default value.
    ///   - decoder: The decoder to use. Defaults to `JSValueDecoder()`.
    /// - Throws: `DecodingError` when the configuration does not match `type`.
    public func decode<T: Decodable>(_ type: T.Type = T.self, decoder: JSValueDecoder = JSValueDecoder()) throws -> T {
        try decoder.decode(type, from: config)
    }
}
